### PR TITLE
ID-42 – on Login, auto-rehash passwords to current preferred cost

### PR DESCRIPTION
**Pull request by NoelLH**

### CODE REVIEW

File: tests/Application/Actions/LoginTest.php [hunk 1]
```diff
@@ -36,6 +36,9 @@ public function testSuccess(): void
             ->shouldBeCalledOnce()
             ->willReturn($personWithHashMatchingRawPasswordFromLoginObject);
 
+        $personRepoProphecy->upgradePasswordIfPossible($person->raw_password, Argument::type(Person::class))
```
Comment (bdsl) [inline tests/Application/Actions/LoginTest.php:39]: We can be specific about which person it should be called for:
```suggestion
        $personRepoProphecy->upgradePasswordIfPossible($person->raw_password, $personWithHashMatchingRawPasswordFromLoginObject)
```

File: src/Repository/PersonRepository.php [hunk 2]
```diff
@@ -89,4 +90,18 @@ public function sendRegisteredEmail(Person $person): bool
     {
         return $this->mailerClient->sendEmail($person->toMailerPayload());
     }
+
+    public function upgradePasswordIfPossible(?string $raw_password, Person $person): void
```
Comment (bdsl) [inline src/Repository/PersonRepository.php:94]: IF we're going to throw let's put the precondition in a docblock:

```suggestion
    /**
     * Generates and persists a new password hash for this person if our existing hash for them wasn't made
     * using our current algorithm and settings.
     *
     * Must only be called for a person who has a password.
     */
    public function upgradePasswordIfPossible(?string $raw_password, Person $person): void
```

**Review by bdsl — COMMENTED** [review]
Added a couple of suggestions, otherwise all LGTM.

File: tests/Application/Actions/LoginTest.php (hunk 1 above)
Comment (NoelLH) [reply tests/Application/Actions/LoginTest.php]: Thanks! I didn't read the existing test properly and had tried the wrong var.

**Review by bdsl — APPROVED** [review]

**Pull request merged**